src/Development/IDE/Core/Shake.hs

```diff
@@ -497,8 +497,15 @@ newSession IdeState{shakeExtras=ShakeExtras{..}, ..} systemActs userActs = do
     let runInShakeSession :: forall a . Action a -> IO (IO a)
         runInShakeSession act = do
               res <- newBarrier
-              let act' = actionCatch @SomeException (Right <$> act) (pure . Left)
-              atomically $ writeTQueue actionQueue (act' >>= liftIO . signalBarrier res)
+              let act' = do
+                    -- work gets reenqueued when the Shake session is restarted
+                    -- it can happen that a work item finished just as it was reenqueud
+                    -- in that case, skipping the work is fine
+                    alreadyDone <- liftIO $ isJust <$> waitBarrierMaybe res
```

Comment:
Just confirming that I get this right: We don’t need to worry about this being racy and the barrier getting updated between our check here and the `unless` since there is only one writer?

Reply:
Only Shake can write, correct. So if this code is running the previous Shake session has already been restarted and there is no one who can fill in the Barrier in between lines 503 and 504